# EPC Analysis
key questions:
- what ratio of records don't have a UPRN
- how can this gap be filled
- what ratio is below E compliance

In [1]:
import pandas as pd
import numpy as np
from dotenv import load_dotenv
import os

In [2]:
load_dotenv()
output_file = os.getenv('EPC_OUTPUT_FILE')
all_epc = pd.read_csv(
    output_file, 
    index_col=0, 
    dtype={'transaction-type': str, 'aircon-present': str}, 
    parse_dates=['lodgement-date', 'lodgement-datetime']
)
# len(all_epc) # 1378713 before any filtering
# remove small properties
all_epc = all_epc[all_epc['floor-area'] >= 40] # m2
dedup = all_epc.sort_values('lodgement-datetime') .drop_duplicates(subset=['building-reference-number'], keep='last')
# df.sort_values('DATE_CHANGED').drop_duplicates('STATION_ID',keep='last')
len(dedup) # 1217573 (1281342 before filter by floor area)

1217573

In [51]:

dedup["uprn"].isna().sum() # 440164

440164

In [42]:
# some of the numbers are from before filter by size
len(dedup[(dedup["lodgement-date"] >= "2014-01-01") & (dedup["uprn"].isna())])  # 282031 - no uprn in the last 10 years - about 1/3

282031

In [36]:
dedup[
    # (dedup["lodgement-date"] >= "2014-01-01") & 
    (dedup["uprn"].isna())
 ].groupby(dedup["lodgement-date"].map(lambda x: x.year)).size()

lodgement-date
2008     9016
2009    38816
2010    24602
2011    26373
2012    36033
2013    28245
2014    25488
2015    25989
2016    24644
2017    27665
2018    33098
2019    32698
2020    21022
2021    23322
2022    31538
2023    40926
2024    13231
dtype: int64

In [43]:
dedup[
    # (dedup["lodgement-date"] >= "2014-01-01") &
    # (dedup["uprn"].isna()) &
    (dedup['asset-rating-band'].isin(['F', 'G']))
    ].groupby(dedup["lodgement-date"].map(lambda x: x.year)).size()

lodgement-date
2008     2400
2009    14680
2010    10458
2011    10706
2012    15649
2013    10740
2014     8716
2015     7866
2016     6828
2017     5253
2018     2848
2019     2136
2020     1383
2021     1322
2022     1297
2023     1175
2024      478
dtype: int64

In [49]:
dedup[
    # (dedup["lodgement-date"] >= "2014-01-01") &
    (dedup["uprn"].isna()) &
    (dedup['asset-rating-band'].isin(['F', 'G']))
][30000:30020]

address1  \
lmk-key                                                                    
98541380302015011616320806300990                                     NaN   
98534190502015011616491150100990                                     NaN   
98555568912015011618234108029816                    GROUND FLOOR OFFICES   
357bd4800943b435ba2a8f9a8f0a1bd8f3edb4a4d20e69d...                   NaN   
95002220202015011914262431900280                                     NaN   
96436380962015011914380813200250                                     NaN   
98603895032015011915303718000897                                     NaN   
8777960642015011916584311600090                                      NaN   
98581580202015011918180521000090                                     NaN   
bfb182b45234523a3a9e7fb8fbe651e79f149eec5ae6491...        STORES BLOCK 2   
98617650062015012013561183990630                                     NaN   
98608825032015012014395413000997                                     NaN   
95819049952015012015315209009467                                     NaN   
98606320942015012108052703800990                                     NaN   
98567515952015012108202604009701                                     NaN   
98679375032015012109521924000998                                     NaN   
98627050202015012110500452200990                                     NaN   
98599230942015012112410041610990                            SECOND FLOOR   
98629860022015012113180435990171                                  UNIT 3   
98446680042015012114195396800090                                     NaN   

                                                                        address2  \
lmk-key                                                                            
98541380302015011616320806300990                                          Unit B   
98534190502015011616491150100990                                         Unit 18   
98555568912015011618234108029816                               57-63 Church Road   
357bd4800943b435ba2a8f9a8f0a1bd8f3edb4a4d20e69d...             Hose Village Hall   
95002220202015011914262431900280                                            RCCG   
96436380962015011914380813200250                         51 Great Ancoats Street   
98603895032015011915303718000897                                  30 High Street   
8777960642015011916584311600090                          Harlow District Council   
98581580202015011918180521000090                    24-30 West Smithfield Street   
bfb182b45234523a3a9e7fb8fbe651e79f149eec5ae6491...     Lamby Way Cleansing Depot   
98617650062015012013561183990630                                     Hair Lounge   
98608825032015012014395413000997                                         Unit 10   
95819049952015012015315209009467                                          Unit 4   
98606320942015012108052703800990                        Truro Motor Services Ltd   
98567515952015012108202604009701                               9 Cross Lane West   
98679375032015012109521924000998                                 1-5 Ault Street   
98627050202015012110500452200990                                         Unit 25   
98599230942015012112410041610990                                   Cadogan House   
98629860022015012113180435990171                                     Grove House   
98446680042015012114195396800090                                     3 Moss Lane   

                                                                         address3  \
lmk-key                                                                             
98541380302015011616320806300990                                     Peacock View   
98534190502015011616491150100990                       Whieldon Industrial Estate   
98555568912015011618234108029816                                              NaN   
357bd4800943b435ba2a8f9a8f0a1bd8f3edb4a4d20e69d...                   2 Harby Lane   
95002220202015011914262431900280

In [44]:
# how many are currently non-compliant
len(dedup[dedup['asset-rating-band'].isin(['F', 'G'])]) # 103,935

103935

In [45]:
# how many with no uprn are non-compliant
len(dedup[
        (dedup["uprn"].isna()) &
        (dedup['asset-rating-band'].isin(['F', 'G']))
    ]) # 41985 (out of 103935)

41985

Observations:
- I don't think we can discount the records before 2014. They are expired, so potentially just as bad, there's plenty in the F and G bands.
- More missing UPRNs in the most recent year 40k, then it drops down to median value 20k-30k from 2022 and before
- There's a lot less in F and G in more recent years. 1175 in 2023, 10740 in 2013
- ~40% of non-compliant properties are missing UPRNs (for all dates)

We will require 4 separate files from DMS:
- EPC list
- os, hmlr, voa, ch data match list
- exemptions (could potentially be included in EPC, but best wait till see actual exemptions data)
- epc - os match list (single record for uprn, multiple for address)

Match algorithm:
- call os api for the 40% with no uprn
- v2 could allow LATS to "secure" the match
- v3 could submit this secured match back to EPC.